### PR TITLE
update query caches on entity updates

### PR DESCRIPTION
**Pull request by sleidig**
Some downloads/exports have been extremely slow because the QueryServices loads all Notes if no date range is defined. This change makes the caching of entities for queries smarter and more reliable in case of changes through database syncing.

### CODE REVIEW

File: src/app/features/reporting/query.service.ts [hunk 1]
```diff
@@ -75,35 +120,25 @@ export class QueryService {
...
+    if (!this.entities || Object.keys(this.entities).length === 0) {
+      for (const entityType of this.entityTypesPrimary) {
+        const loadedEntities = await this.entityMapper.loadType(entityType);
+        this.setEntities(entityType, loadedEntities);
+      }
     }
 
-    await Promise.all(dataPromises);
+    if (
+      this.entityTypesTimebound.some((t) => query.includes(t[0].ENTITY_TYPE)) &&
```
Comment (sleidig) [inline src/app/features/reporting/query.service.ts:132]: same filter could be applied to "primary" entities as well

**Comment by sleidig** [comment]
TODO: unit tests?

File: src/app/features/reporting/query.service.ts [hunk 2]
```diff
@@ -20,14 +21,59 @@ const jsonQuery = require("json-query");
...
+      ...this.entityTypesTimebound.map((e) => e[0]),
+    ]) {
+      this.entityMapper
+        .receiveUpdates(entityType)
+        .subscribe((u) => this.updateCachedData(u));
+    }
+  }
+
+  private updateCachedData(u: UpdatedEntity<Entity>) {
+    const cachedEntities = this.entities[u.entity.getConstructor().ENTITY_TYPE];
```
Comment (TheSlimvReal) [inline src/app/features/reporting/query.service.ts:65]:
```suggestion
    const cachedEntities = this.entities[u.entity.getType()];
```

File: src/app/features/reporting/query.service.ts (hunk 1 above)
Comment (TheSlimvReal) [reply src/app/features/reporting/query.service.ts:132]: But its a bit more challenging as we would also need to check for some (at least one) of the more specific functions: e.g. calling `getAttendanceArray(true)` needs the `ChildSchoolRelation` to be cached.

**Review by TheSlimvReal — COMMENTED** [review]
Looks quite good already we might have to split up the query server and this entity cache though as its getting quite complex. (Maybe even with the `MockEntityMapper` which also allows to request entities synchronously).

**Comment by sleidig** [comment]
also see #1697

**Comment by TheSlimvReal** [comment]
Part of #1711

**Pull request closed without merging**